File: bindings/lua/Polycode/Vector2.lua
class "Vector2"


function Vector2:__getvar(name)
	if name == "x" then
		return Polycode.Vector2_get_x(self.__ptr)
	elseif name == "y" then
		return Polycode.Vector2_get_y(self.__ptr)
	end
end

function Vector2:__setvar(name,value)
	if name == "x" then
		Polycode.Vector2_set_x(self.__ptr, value)
		return true
	elseif name == "y" then
		Polycode.Vector2_set_y(self.__ptr, value)
		return true
	end
	return false
end
function Vector2:Vector2(...)
	local arg = {...}
	for k,v in pairs(arg) do
		if type(v) == "table" then
			if v.__ptr ~= nil then
				arg[k] = v.__ptr
			end
		end
	end
	if self.__ptr == nil and arg[1] ~= "__skip_ptr__" then
		self.__ptr = Polycode.Vector2(unpack(arg))
	end
end

function Vector2:set(x, y)
	local retVal = Polycode.Vector2_set(self.__ptr, x, y)
end

function Vector2:distance(vec2)
	local retVal = Polycode.Vector2_distance(self.__ptr, vec2.__ptr)
	return retVal
end

function Vector2:length()
	local retVal =  Polycode.Vector2_length(self.__ptr)
	return retVal
end

function Vector2:dot(u)
	local retVal = Polycode.Vector2_dot(self.__ptr, u.__ptr)
	return retVal
end

function Vector2:crossProduct(vec2)
	local retVal = Polycode.Vector2_crossProduct(self.__ptr, vec2.__ptr)
	return retVal
end

function Vector2:angle(vec2)
	local retVal = Polycode.Vector2_angle(self.__ptr, vec2.__ptr)
	return retVal
end

function Vector2:Normalize()
	local retVal =  Polycode.Vector2_Normalize(self.__ptr)
end

function Vector2:__delete()
	if self then Polycode.delete_Vector2(self.__ptr) end
end

File: bindings/javascript/Polycode/Vector2.js
function Vector2(x,y) {
	if(arguments[0] != "__skip_ptr__") {
		this.__ptr = Polycode.Vector2(x,y)
	}
	Object.defineProperties(this, {
		'x': { enumerable: true, configurable: true, get: Vector2.prototype.__get_x, set: Vector2.prototype.__set_x},
		'y': { enumerable: true, configurable: true, get: Vector2.prototype.__get_y, set: Vector2.prototype.__set_y}
	})
}


Vector2.prototype.__get_x = function() {
	return Polycode.Vector2__get_x(this.__ptr)
}

Vector2.prototype.__set_x = function(val) {
	Polycode.Vector2__set_x(this.__ptr, val)
}

Vector2.prototype.__get_y = function() {
	return Polycode.Vector2__get_y(this.__ptr)
}

Vector2.prototype.__set_y = function(val) {
	Polycode.Vector2__set_y(this.__ptr, val)
}

Duktape.fin(Vector2.prototype, function (x) {
	if (x === Vector2.prototype) {
		return;
	}
	Polycode.Vector2__delete(x.__ptr)
})

Vector2.prototype.set = function(x,y) {
	Polycode.Vector2_set(this.__ptr, x, y)
}

Vector2.prototype.distance = function(vec2) {
	return Polycode.Vector2_distance(this.__ptr, vec2)
}

Vector2.prototype.length = function() {
	return Polycode.Vector2_length(this.__ptr)
}

Vector2.prototype.dot = function(u) {
	return Polycode.Vector2_dot(this.__ptr, u)
}

Vector2.prototype.crossProduct = function(vec2) {
	return Polycode.Vector2_crossProduct(this.__ptr, vec2)
}

Vector2.prototype.angle = function(vec2) {
	return Polycode.Vector2_angle(this.__ptr, vec2)
}

Vector2.prototype.Normalize = function() {
	Polycode.Vector2_Normalize(this.__ptr)
}
